frontend-web/src/coffee-call/call-aggregation.ts

```diff
@@ -0,0 +1,71 @@
+import { delay, map, merge, Observable, of, scan, switchMap, tap } from "rxjs";
+import { ReceivedCallMessage } from "./protocol";
+
+export const enum CoffeeCallState {
+  inactive,
+  announced,
+  inProgress,
+  canceled,
+};
+
+export class CoffeeCall {
+  state: CoffeeCallState = CoffeeCallState.inactive;
+  participants: Set<string> = new Set();
+
+  constructor(public messages: ReceivedCallMessage[]) {
+    if (messages.length === 0) {
+      return;
+    }
+
+    for (const message of messages) {
+      switch (message.type) {
+        case 'start':
+          this.state = CoffeeCallState.inProgress;
+          this.participants.add(message.name);
+          break;
+        case 'join':
+          if (this.state !== CoffeeCallState.inProgress) {
+            this.state = CoffeeCallState.announced;
+          }
+          this.participants.add(message.name);
+          break;
+        case 'leave':
+          this.participants.delete(message.name);
+          break;
+        default:
+          assertUnreachable(message.type);
+      }
+    }
+
+    if (this.participants.size === 0) {
+      this.state = CoffeeCallState.canceled;
+    }
+  }
+}
+
+export function aggregatedCoffeeCall(messages: Observable<ReceivedCallMessage>): Observable<CoffeeCall> {
+  const reset = 'RESET' as const;
+  type Reset = typeof reset;
+  const accumulateMessagesUntilReset = (oldMessages: ReceivedCallMessage[], newMessage: ReceivedCallMessage | Reset) =>
+    newMessage === reset ? [] : [...oldMessages, newMessage];
+
+  const accumulateMessagesWindowInSeconds = 3 * 60 * 1000;
+
+  const accumulatedMessagesInTimeWindow = messages.pipe(
+    switchMap((message) => {
+      const forwardedMessage = of(message);
+      const resetSignal = of(reset).pipe(delay(accumulateMessagesWindowInSeconds));
+      return merge(forwardedMessage, resetSignal);
+    }),
+    scan(accumulateMessagesUntilReset, [] as ReceivedCallMessage[])
+  );
+
+  return accumulatedMessagesInTimeWindow.pipe(
+    map((messageList) => new CoffeeCall(messageList)) // Todo: this creates a new call for every message
+  );
+}
+
+
+function assertUnreachable(x: never): never {
```

Comment:
[nit] either remove the `x` arg and param in l. 36, or make some actual use of it (e.g. include in the Error)

Reply:
Can't remove the parameter, because it is needed so that TS can statically check that the value is `never`. I'll include it in the error message, although if everything goes to plan that code should be dead.

Reply:
???

I don't get any TS errors when the param is removed.

Reply:
You lose this nifty error hint (notice I removed the `"leave"` type, which means that typescript tries to assign it to `never`, which causes it to fail at compile time, effectively guarding against missing a case-statement when adding a new message type):

![image](https://user-images.githubusercontent.com/1775745/190648085-5a6dd701-a882-4f26-8b29-6209b354dcf4.png)